src/main/java/com/opentable/kafka/util/LogProgressRestoreListener.java

```diff
@@ -0,0 +1,144 @@
+package com.opentable.kafka.util;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.IntStream;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.KafkaStreams.State;
+import org.apache.kafka.streams.KafkaStreams.StateListener;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LogProgressRestoreListener implements StateRestoreListener, StateListener {
+    private static final Logger LOG = LoggerFactory.getLogger(LogProgressRestoreListener.class);
+
+    @GuardedBy("this")
+    private final Map<String, Partitions> restoreState = new LinkedHashMap<>();
+    private Instant lastPrint = Instant.now(), restoreStart = Instant.now();
+
+    private final StateListener delegate;
+
+    public LogProgressRestoreListener() {
+        this(null);
```

Comment:
@VisibleForTesting?

Reply:
it's used if you don't have a further restore listener (only one can be installed, so if you want progress as well as some other behavior, this allows that).